#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/supermarket_sales.csv')

In [4]:
df[:2]

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   

             Product line  Unit price  Quantity   Tax 5%     Total      Date  \
0       Health and beauty       74.69         7  26.1415  548.9715  1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200  3/8/2019   

    Time  Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08  Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29     Cash   76.40                 4.761905        3.8200     9.6

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [5]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

Найприбутковішим є філіал `C` з сумарним прибутком 5265.1765.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [8]:
sum_gross_income = df.pivot_table(index='Branch',
                                  columns='City',
                                  values='gross income',
                                  aggfunc='sum')

In [9]:
sum_gross_income

City    Mandalay  Naypyitaw     Yangon
Branch                                
A            NaN        NaN  5057.1605
B       5057.032        NaN        NaN
C            NaN  5265.1765        NaN

In [10]:
city_with_max_gross = sum_gross_income.max().idxmax()
print(f"Філіал з найбільшим прибутком знаходиться в місті {city_with_max_gross}")

Філіал з найбільшим прибутком знаходиться в місті Naypyitaw


In [11]:
# Другий спосіб знайти яке місто відповідає філії C
df[df.Branch == 'C'].City.unique()

array(['Naypyitaw'], dtype=object)

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [12]:
invoices_by_category = df.pivot_table(index='Branch',
                                     columns='Product line',
                                     values='Invoice ID',
                                     aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [13]:
top_branch = invoices_by_category['Electronic accessories'].idxmax()
print(f"Найбільше інвойсів у категорії 'Electronic accessories' має філія: {top_branch}")

Найбільше інвойсів у категорії 'Electronic accessories' має філія: A


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


## Питання 1
Чи взаємопов’язана середня оцінка задоволення (Rating) із категорією товару по містам?  
Мета: з’ясувати, в яких категоріях товарів клієнти залишають вищі оцінки.

In [14]:
# Середній рейтинг по Product line по містам
pd.pivot_table(df, index='Product line', columns='City', values='Rating', aggfunc='mean').round(2)

City                    Mandalay  Naypyitaw  Yangon
Product line                                       
Electronic accessories      7.12       6.75    6.91
Fashion accessories         6.72       7.44    6.88
Food and beverages          6.99       7.08    7.25
Health and beauty           7.10       7.00    6.90
Home and lifestyle          6.52       7.06    6.93
Sports and travel           6.51       7.03    7.26

### Висновок:   
В місті `Mandalay` по категоріям товарів `Home and lifestyle`, `Sports and travel` найнижчі середні значення рейтингу серед всіх міст. По низьким показникам можна сказати, що рівень сервісу в магазині потребує змін, або мала кількість консультантів, що допомагають зробити вибір, або не дуже великий асортимент товару і клієнти цим незадоволені.  

### Рекомендації:
- Переглянути асортимент товару в категоріях, де низький рейтинг.
- Проаналізувати рівнь сервісу по катеріям.
- Акцентувати маркетинг на найкращих поєднаннях рейтинг-категорія.

## Питання 2:
У яких містах і в які дні тижня купують найбільше одиниць конкретної категорії товарів?  
Мета: зрозуміти найзатребуваніші дні + географію попиту по категоріях.


In [15]:
# Додамо колонку з днем тижня
df['Weekday'] = pd.to_datetime(df['Date']).dt.day_name()

In [17]:
df[:2]

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   

             Product line  Unit price  Quantity   Tax 5%     Total      Date  \
0       Health and beauty       74.69         7  26.1415  548.9715  1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200  3/8/2019   

    Time  Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08  Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29     Cash   76.40                 4.761905        3.8200     9.6   

    Weekday  
0  Saturday  
1    Friday

In [18]:
# Кількість одиниць товарів по містах і днях тижня
product_line_by_weekday = pd.pivot_table(df, index=['Product line', 'City'], columns='Weekday', values='Quantity', aggfunc='sum')

In [19]:
# Задаємо бажаний порядок днів тижня
ordered_days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

# Перевпорядковуємо стовпці відповідно до списку
product_line_by_weekday = product_line_by_weekday[ordered_days]
product_line_by_weekday

Weekday                           Monday  Tuesday  Wednesday  Thursday  \
Product line           City                                              
Electronic accessories Mandalay       39       24         41        84   
                       Naypyitaw      36       58         12        63   
                       Yangon         49       69         36        66   
Fashion accessories    Mandalay       24       39         47        45   
                       Naypyitaw      25       42         65        51   
                       Yangon         43       38         22        23   
Food and beverages     Mandalay       35       70         48        20   
                       Naypyitaw      40       84         64        54   
                       Yangon         23       14         73        42   
Health and beauty      Mandalay       50       72         60        25   
                       Naypyitaw      44       54         28        22   
                       Yangon         46       29         43        49   
Home and lifestyle     Mandalay       25       56         16        35   
                       Naypyitaw      30       30         57        21   
                       Yangon         56       50         29        39   
Sports and travel      Mandalay       21       50         74        45   
                       Naypyitaw      16       38         47        31   
                       Yangon         36       45         38        40   

Weekday                           Friday  Saturday  Sunday  
Product line           City                                 
Electronic accessories Mandalay       17        87      24  
                       Naypyitaw      48        30      86  
                       Yangon         36        42      24  
Fashion accessories    Mandalay       65        66      11  
                       Naypyitaw      53        64      42  
                       Yangon         38        49      50  
Food and beverages     Mandalay       35        27      35  
                       Naypyitaw      24        56      47  
                       Yangon         36        68      57  
Health and beauty      Mandalay       59        36      18  
                       Naypyitaw      50        49      30  
                       Yangon         36        22      32  
Home and lifestyle     Mandalay       33        73      57  
                       Naypyitaw      19        33      55  
                       Yangon         79        43      75  
Sports and travel      Mandalay       46        46      40  
                       Naypyitaw      35        49      49  
                       Yangon         49        79      46

In [20]:
# Мінімальна кількість проданих одиниць у кожній комбінації Product line + City
min_values = product_line_by_weekday.min(axis=1)

# День тижня, коли це мінімальне значення трапилось
min_days = product_line_by_weekday.idxmin(axis=1)

# Об’єднуємо у підсумкову таблицю
min_summary = pd.DataFrame({
    "Мінімальна кількість": min_values,
    "День тижня": min_days
    })

min_summary

Мінімальна кількість День тижня
Product line           City                                      
Electronic accessories Mandalay                     17     Friday
                       Naypyitaw                    12  Wednesday
                       Yangon                       24     Sunday
Fashion accessories    Mandalay                     11     Sunday
                       Naypyitaw                    25     Monday
                       Yangon                       22  Wednesday
Food and beverages     Mandalay                     20   Thursday
                       Naypyitaw                    24     Friday
                       Yangon                       14    Tuesday
Health and beauty      Mandalay                     18     Sunday
                       Naypyitaw                    22   Thursday
                       Yangon                       22   Saturday
Home and lifestyle     Mandalay                     16  Wednesday
                       Naypyitaw                    19     Friday
                       Yangon                       29  Wednesday
Sports and travel      Mandalay                     21     Monday
                       Naypyitaw                    16     Monday
                       Yangon                       36     Monday

### Висновки:  
На основі зведеної таблиці `product_line_by_weekday` легко аналізувати в які дні тижня і які саме товари користуються найменшим попитом, тут доцільно буде побудувати теплову карту. На основі цієї таблиці побудована таблиця `min_summary`, в якій вказано найменше значення для кожної комбінації (Product line, City) і у який саме день тижня це було.

### Рекомендації:
- Оптимізувати запаси та змінити розкладку продуктів/товарів у залежності від дня.  
- Запустити таргентовані акції у містах/днях, де попит слабший.

## Питання 3
Чи впливає тип клієнта (Customer type) і час доби на середній чек?  
Мета: дослідити чи схильні члени клубу витрачати більше в певні години, ніж звичайні покупці?

In [21]:
# Обчислимо середній чек (Total) з розбивкою по містах і типу клієнта (Member/Normal).
table_avg_check = df.pivot_table(index='Customer type',
                                          columns='City',
                                          values='Total',
                                          aggfunc='mean')

In [22]:
table_avg_check.round(1)

City           Mandalay  Naypyitaw  Yangon
Customer type                             
Member            325.5      336.6   321.2
Normal            314.3      337.7   303.8

In [23]:
# Функція, що з часу в форматі "година:хвилина" дістає тільки години, і конвертує їх у тип int
def get_hour(time_str):
    hour_str = time_str.split(":")[0]
    return int(hour_str)

In [24]:
# Додамо колонку з годиною покупки
df['Hour'] = df['Time'].apply(get_hour)

In [25]:
df[:2]

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   

             Product line  Unit price  Quantity   Tax 5%     Total      Date  \
0       Health and beauty       74.69         7  26.1415  548.9715  1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200  3/8/2019   

    Time  Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08  Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29     Cash   76.40                 4.761905        3.8200     9.6   

    Weekday  Hour  
0  Saturday    13  
1    Friday    10

In [26]:
# Обчислимо середній чек (Total) за годинами (Hour) з розбивкою по містах і типу клієнта (Member/Normal).
table_avg_check_by_hour = df.pivot_table(index='Hour',
                                          columns=['City', 'Customer type'],
                                          values='Total',
                                          aggfunc= 'mean')

In [27]:
table_avg_check_by_hour.round(1)

City          Mandalay        Naypyitaw        Yangon       
Customer type   Member Normal    Member Normal Member Normal
Hour                                                        
10               342.7  340.4     319.6  294.4  242.1  337.8
11               334.4  297.5     365.2  411.7  305.3  342.2
12               330.0  350.5     235.3  283.0  323.1  244.6
13               300.2  294.0     370.6  399.7  291.7  369.5
14               427.4  340.7     388.3  346.1  373.3  313.2
15               320.3  319.6     295.6  290.3  393.9  210.5
16               261.3  221.5     320.9  416.9  264.4  391.2
17               360.0  424.1     305.8  247.8  417.1  278.5
18               262.2  280.2     302.4  424.7  225.4  226.0
19               316.8  336.1     350.0  381.7  399.2  367.2
20               307.4  266.8     454.6  269.9  329.2  206.8

### Висновки:  
- У місті Mandalay:
  -- Члени клубу мають вищі середні чеки у період 14.00-17.00 (піки до 427 грошових одиниць)
  -- Звичайні клієнти особливо активні з 17 до 19 (трохи вище 424 грошових одиниць)
- У місті Naypyitaw:
  -- Карткові клієнти теж показують високі суми чека між 13:00 і 15:00.
  -- Рекордно високий середній чек 454.6 о 20 — це знову Member-клієнти, отже вечір у Naypyitaw — ідеальний період для залучення лояльних покупців.
- У місті Yangon:
  -- Чіткий пік для клієнтів з карткою — о 17 (417 грошових одиниць) та о 19 (399 грошових одиниць)
  -- Унікальність: тут "Members" купують значно більше ввечері, ніж "Normal".
  
У всіх містах є годинні слоти, коли члени клубу (Members) витрачають суттєво більше — переважно в обід та ввечері (13–17 і 19–20).  
Клієнти без картки показують вищі чеки в окремі пікові години (наприклад, о 16 і 18 у Naypyitaw).


### Рекомендації:
- Запуск таргетованих вечірніх акцій для лояльних клієнтів у Yangon і Naypyitaw (наприклад, бонуси після 17:00).  
- Залучення звичайних покупців у Mandalay через стимули в години, де вони не активні (наприклад, обідні знижки).